edalize/libero.py

```diff
@@ -0,0 +1,181 @@
+                shouldExit = 1
+        if shouldExit:
+            exit(1)
+
+    """ Configuration is the first phase of the build
+
+    This writes the project TCL file. It first collects all
+    sources, IPs and constraints and then writes them to the TCL file along
+    with the build steps.
+    """
+
+    def configure_main(self):
+        self._set_tool_options_defaults()
+        self._check_mandatory_options()
+        (src_files, incdirs) = self._get_fileset_files(force_slash=True)
+        self.jinja_env.filters['src_file_filter'] = self.src_file_filter
+        self.jinja_env.filters['constraint_file_filter'] = self.constraint_file_filter
+
+        escaped_name = self.name.replace(".", "_")
+        synth_opts = self._get_tool_params(
+            self.tool_options['synth_options']) if 'synth_options' in self.tool_options else ''
+        pnr_opts = self._get_tool_params(
+            self.tool_options['pnr_options']) if 'pnr_options' in self.tool_options else ''
+
+        template_vars = {
+            'name': escaped_name,
+            'src_files': src_files,
+            'incdirs': incdirs,
+            'vlogparam': self.vlogparam,
+            'vlogdefine': self.vlogdefine,
+            'tool_options': self.tool_options,
+            'synth_options': synth_opts,
+            'pnr_options': pnr_opts,
+            'toplevel': self.toplevel,
+            'generic': self.generic,
+            'prj_root': "./prj",
+            'op': "{",
+            'cl': "}",
+            'sp': " "
+        }
+
+        # Set preferred HDL language based on file type amount if not user defined.
+        # According to docs, projects can be mixed but one language must be
+        # defined as preferred
+        if not "hdl" in self.tool_options:
+            verilogFiles = 0
+            VHDLFiles = 0
+            for f in src_files:
+                t = f.file_type.split('-')[0]
+                if t == "verilogSource" or t == "systemVerilogSource":
+                    verilogFiles += 1
+                elif t == "vhdlSource":
+                    VHDLFiles += 1
+            if verilogFiles >= VHDLFiles:
+                self.tool_options["hdl"] = "VERILOG"
+            else:
+                self.tool_options["hdl"] = "VHDL"
+
+        # Render the TCL project file
+        self.render_template('libero-project.tcl.j2',
+                             escaped_name + '-project.tcl',
+                             template_vars)
+
+        # Render the TCL run file
+        self.render_template('libero-run.tcl.j2',
+                             escaped_name + '-run.tcl',
+                             template_vars)
+
+        # Render the Synthesize TCL file
+        self.render_template('libero-syn-user.tcl.j2',
+                             escaped_name + '-syn-user.tcl',
+                             template_vars)
+
+        logger.info("Cores and Libero TCL Scripts generated.")
+
+    def src_file_filter(self, f):
```

Comment:
In this function we need to check vhdl (and verilog?) files if they have `logical_name`set to anything. For those files we need to add `-library {<logical_name>}` to the generated tcl

Reply:
I saw how it's done on Vivado but on Libero docs there is no "-library" setting to import files.

Here's the import_files doc:

![image](https://user-images.githubusercontent.com/20382/107394684-e6a63880-6ada-11eb-9e4a-616cf5fe3a33.png)

Reply:
For this, I'll have to iterate thru all "logical_name"s and import each since the expected script is:

```
import_files \
         -library {work2} \
         -hdl_source {file.vhdl} 
```

Reply:
Did it, it would become something like:

```
...
# Import HDL sources and constraints
import_files \
        -sdc {sdc_file} \
        -hdl_source {sv_file.sv} \
        -hdl_source {vlog_file.v} \
        -hdl_source {vlog05_file.v} \
        -hdl_source {vhdl_file.vhd} \
        -hdl_source {vhdl2008_file} \
        -hdl_source {another_sv_file.sv} \
        -io_pdc {pdc_constraint_file.pdc} \

# Import HDL sources on specifit libraries (logical_names)
import_files \
        -library {libx} \
        -hdl_source {vhdl_lfile} \
        -hdl_source {vhdl_lfile2} \

import_files \
        -library {libxy} \
        -hdl_source {vhdl_lfile3} \
...
```